crates/matrix-sdk/src/event_handler.rs

```diff
@@ -343,85 +355,85 @@ impl Client {
+
+            self.call_event_handlers(room, raw_event, event_kind, &event_type, encryption_info)
+                .await;
+        }
 
         Ok(())
     }
 
-    async fn handle_sync_events_wrapped_with<'a, T: 'a, U: 'a>(
+    async fn call_event_handlers(
         &self,
         room: &Option<room::Room>,
-        list: &'a [U],
-        get_event_details: impl Fn(&'a U) -> (&'a Raw<T>, Option<&'a EncryptionInfo>),
-        get_id: impl Fn(&Raw<T>) -> serde_json::Result<(EventKind, Cow<'_, str>)>,
-    ) -> serde_json::Result<()> {
-        for x in list {
-            let (raw_event, encryption_info) = get_event_details(x);
-            let (ev_kind, ev_type) = get_id(raw_event)?;
-            let event_handler_id = (ev_kind, &*ev_type);
-
-            // Construct event handler futures
-            let futures: Vec<_> = self
-                .event_handlers()
-                .await
-                .get(&event_handler_id)
+        raw: &RawJsonValue,
+        ev_kind: EventKind,
+        ev_type: &str,
+        encryption_info: Option<&EncryptionInfo>,
+    ) {
+        // Construct event handler futures
+        let futures: Vec<_> = {
+            let handlers_lock = self.event_handlers().await;
+
+            let non_room_event_handlers = handlers_lock
+                .get(&EventHandlerKey { ev_kind, ev_type, room_id: None })
                 .into_iter()
-                .flatten()
+                .flatten();
+
+            let room_event_handlers = room.iter().flat_map(|r| {
+                let room_id = Some(r.room_id().to_owned());
+                handlers_lock
+                    .get(&EventHandlerKey { ev_kind, ev_type, room_id })
+                    .into_iter()
+                    .flatten()
+            });
+
+            non_room_event_handlers
+                .chain(room_event_handlers)
                 .map(|handler_wrapper| {
-                    let data = EventHandlerData {
-                        client: self.clone(),
-                        room: room.clone(),
-                        raw: raw_event.json(),
-                        encryption_info,
-                        handle: handler_wrapper.handle,
-                    };
+                    let client = self.clone();
+                    let room = room.clone();
+                    let handle = handler_wrapper.handle;
+                    let data = EventHandlerData { client, room, raw, encryption_info, handle };
+
                     (handler_wrapper.handler_fn)(data)
                 })
-                .collect();
+                .collect()
+        };
 
-            // Run the event handler futures with the `self.event_handlers` lock
-            // no longer being held, in order.
-            for fut in futures {
-                fut.await;
-            }
+        // Run the event handler futures with the `self.event_handlers` lock
+        // no longer being held, in order.
+        for fut in futures {
+            fut.await;
```

Comment:
I know not technically part of this PR BUT why are not executing them in parallel? Do we promise somewhere that they are run exactly in the order they are added? Do we think that is a guarantee we want to make? Why not do [`futures::join_all!(futures)?`](https://docs.rs/futures/latest/futures/future/fn.join_all.html) on them instead? That would parallelize them and potentially improve performance..

Sure, in that case we should do that on `non_room_event_handlers` and then on `room_event_handlers`, because I think the promise that those global once are run first and the others second, kinda does make sense (though it isn't really said anywhere, which it probably should be).

Reply:
I think it's not really common to have multiple event handlers that fire on the same event. Either way, let's discuss this on Matrix, or in a meeting.